snapcraft/remote/launchpad.py

```diff
@@ -0,0 +1,478 @@
+            self._download_log(build)
+
+    def _get_builds_collection_entry(self, snap: Entry) -> Optional[Entry]:
+        logger.debug("Fetching builds collection information from Launchpad...")
+        url = snap.builds_collection_link
+        return self._lp_load_url(url)
+
+    def _get_builds(self, snap: Entry) -> List[Dict[str, Any]]:
+        builds_collection = self._get_builds_collection_entry(snap)
+        if builds_collection is None:
+            return []
+
+        return builds_collection.entries
+
+    def _get_snap(self) -> Optional[Entry]:
+        try:
+            return self._lp.snaps.getByName(name=self._lp_name, owner=self._lp_owner)
+        except restfulclient.errors.NotFound:
+            return None
+
+    def _issue_build_request(self, snap: Entry) -> Entry:
+        dist = self._lp.distributions["ubuntu"]
+        archive = dist.main_archive
+        return snap.requestBuilds(
+            archive=archive,
+            pocket="Updates",
+        )
+
+    def _lp_load_url(self, url: str) -> Entry:
+        """Load Launchpad url with a retry in case the connection is lost."""
+        try:
+            return self._lp.load(url)
+        except ConnectionResetError:
+            self._lp = self.login()
+            return self._lp.load(url)
+
+    def _wait_for_build_request_acceptance(self, build_request: Entry) -> None:
+        # Not be be confused with the actual build(s), this is
+        # ensuring that Launchpad accepts the build request.
+        while build_request.status == "Pending":
+            # Check to see if we've run out of time.
+            self._check_timeout_deadline()
+
+            logger.debug("Waiting on Launchpad build request...")
+            logger.debug(
+                "status=%s error=%s", build_request.status, build_request.error_message
+            )
+
+            time.sleep(1)
+
+            # Refresh status.
+            build_request.lp_refresh()
+
+        if build_request.status == "Failed":
+            # Build request failed.
+            self.cleanup()
+            raise errors.RemoteBuildError(build_request.error_message)
+
+        if build_request.status != "Completed":
+            # Shouldn't end up here.
+            self.cleanup()
+            raise errors.RemoteBuildError(
+                f"Unknown builder error - reported status: {build_request.status}"
+            )
+
+        if not build_request.builds.entries:
+            # Shouldn't end up here either.
+            self.cleanup()
+            raise errors.RemoteBuildError(
+                "Unknown builder error - no build entries found."
+            )
+
+        build_number = _get_url_basename(build_request.self_link)
+        logger.debug("Build request accepted: %s", build_number)
+
+    def login(self) -> Launchpad:
+        """Login to launchpad."""
+        try:
+            return Launchpad.login_with(
+                f"{self._app_name} remote-build",
+                "production",
```

Comment:
This might be worth making an optional parameter that defaults to `"production"`. For certain integration/e2e tests we might want to talk to staging if the test could make a mess on launchpad.

Reply:
Agreed, I'm going to leave it as-is and we can add this in the a future PR.